File: backend/transcripts/schemas/transcript.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from models.enums import TranscriptSourceType, TranscriptStatus, TranscriptStreamType


class TranscriptCreate(BaseModel):
    workflow_id: UUID | None = None
    call_id: UUID | None = None
    lead_id: UUID | None = None
    source_type: TranscriptSourceType = TranscriptSourceType.call


class TranscriptChunkCreate(BaseModel):
    speaker: str
    text: str
    stream_type: TranscriptStreamType = TranscriptStreamType.user
    chunk_index: int | None = None


class TranscriptResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    workflow_id: UUID | None
    call_id: UUID | None
    lead_id: UUID | None
    transcript_status: TranscriptStatus
    source_type: TranscriptSourceType
    duration_seconds: int | None
    created_at: datetime
    updated_at: datetime


class TranscriptChunkResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    transcript_id: UUID
    chunk_index: int
    speaker: str
    text: str
    stream_type: TranscriptStreamType
    created_at: datetime


class TranscriptEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    transcript_id: UUID
    event_type: str
    sequence: int
    payload: dict
    created_at: datetime

File: frontend/types/transcript.ts
export type TranscriptStatus =
  | 'created'
  | 'active'
  | 'paused'
  | 'completed'
  | 'failed'
  | 'archived'

export type TranscriptSourceType = 'call' | 'upload' | 'realtime' | 'manual'
export type TranscriptStreamType = 'agent' | 'user' | 'system' | 'mixed'

export interface Transcript {
  id: string
  workflow_id: string | null
  call_id: string | null
  lead_id: string | null
  transcript_status: TranscriptStatus
  source_type: TranscriptSourceType
  duration_seconds: number | null
  created_at: string
  updated_at: string
}

export interface TranscriptChunk {
  id: string
  transcript_id: string
  chunk_index: number
  speaker: string
  text: string
  stream_type: TranscriptStreamType
  created_at: string
}

export interface TranscriptEvent {
  id: string
  transcript_id: string
  event_type: string
  sequence: number
  payload: Record<string, unknown>
  created_at: string
}

export interface TranscriptReplay {
  transcript_id: string
  replayed_events: number
  final_status: TranscriptStatus
  chunk_count: number
}

export interface ChunksPage {
  items: TranscriptChunk[]
  total: number
  page: number
  page_size: number
}
